add CRUD operations for Flask, change backend to non-interactive for Flask to display the statistics as a png

yavvorn opened the pull request:
add CRUD operations for Flask, change backend to non-interactive for Flask to display the statistics as a png

yavvorn committed:
add CRUD operations for Flask, change backend to non-interactive for Flask to display the statistics as a png

slavovthinks committed:
Refactor code

yavvorn committed:
connect CRUD operations to Expense_tracker DB, add an html file and a styles.css file, add sample ajax for later use

yavvorn committed:
remove unnecessary png file

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -1,92 +1,163 @@
-
+db_pool = pool.SimpleConnectionPool(1, 20, database="Finance_tracker",
+                                    user="postgres",
+                                    password="pass1234!",
```
Move credentials to env vars

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -1,92 +1,163 @@
 app = Flask(__name__)
 
 ALLOWED_CATEGORIES = ['Entertainment', 'Bills', 'Food', 'Cash', 'Necessities']
 PATH_TO_CSV_FILE = '../spending.csv'
```
remove old constants

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -1,92 +1,163 @@
+#     return expense
 
 
-@app.route('/expenses', methods=['POST'])
```
Return CRUD endpoints

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -1,92 +1,163 @@
 
-@app.route('/expenses', methods=['POST'])
+@app.route('/create', methods=['POST'])
 def create_expense():
```
create expense can accept expense dict and create an expense in the database

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -1,92 +1,163 @@
-def get_statistics():
+    except Exception as e:
+        conn.rollback()  # Rollback the transaction in case of error
+        return str(e), 400  # Return error message with HTTP status code 400 (Bad Request)
```
Errors should be in json format

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -1,92 +1,163 @@
-            overall_usage[expense['expense']] += expense['value']
-
-    return overall_usage
+    conn = get_db()
```
extract query execution logic in execute_query function

slavovthinks commented on pyproject.toml:
```diff
@@ -9,6 +9,8 @@ readme = "README.md"
 python = "3.10.*"
```
Create initial DDL/migration/DB structure script.

```sql
CREATE TABLE blabla...
```

slavovthinks committed:
CRUD is back babyy

yavvorn committed:
CRUD is back baby
assistant funcs for connection and sql execution added
Global var for db pass added

yavvorn committed:
Merge remote-tracking branch 'origin/flask_additions' into flask_additions

# Conflicts:
#	spending_tracker/app.py

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -0,0 +1,151 @@
+    cur = None
+    try:
+        cur = conn.cursor()
+    except Exception as e:
```
If it fails it fails dont handle the error here

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -0,0 +1,151 @@
+            cur.execute(query)
+        data = cur.fetchall()
+    except TypeError as e:
+        return {"error": str(e)}, 400
```
Move http specific logic in the controller/route

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -0,0 +1,151 @@
+            cur.execute(query, args)
+        else:
+            cur.execute(query)
+    except TypeError as e:
```
HTTP -> Route

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -0,0 +1,151 @@
+    """
+    Returns all the expenses from the database.
+    """
+    cur, conn = create_cursor()
```
cur in query executor.

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -0,0 +1,151 @@
+    expense_value = data.get('value')
+    expense_id = data.get("id")
+
+    cur, conn = create_cursor()
```
c -> q

slavovthinks commented on spending_tracker/config.py:
```diff
@@ -0,0 +1,10 @@
+db_pool = pool.SimpleConnectionPool(
+    1, 20,
+    database="Finance_tracker",
+    user="postgres",
```
ENV VARS

yavvorn committed:
query_executor updated alongside other funcs
env vars introduced via poetry
.env.example added

yavvorn committed:
Added DDL for the DB table.

yavvorn committed:
Added DDL for the DB table. try-except for expense creation, added tests but encountered an issue which is explained in test_app.py lines 43-47

yavvorn committed:
Added DDL for the DB table. try-except for expense creation, added tests but encountered an issue which is explained in test_app.py lines 43-47

slavovthinks commented on init_database.py:
```diff
@@ -0,0 +1,18 @@
+
+
+if __name__ == "__main__":
+    with app.app_context(): #ensures Flask context is available so that g is accessible
```
Comment style should be the following way`{space}{space}# {blabla}`

slavovthinks commented on tests/test_app.py:
```diff
@@ -0,0 +1,83 @@
+    expense_id = 2
+    response = app.test_client().get(f"/expenses/{expense_id}")
+    assert response.status_code == 200
+    assert response.json == {"Food": 50}
```
you need to mock db response

slavovthinks commented on tests/test_app.py:
```diff
@@ -0,0 +1,83 @@
+
+
+def test_get_expense_non_existent():
+    expense_id = sys.maxsize
```
use db mocks

slavovthinks committed:
WIP: UT fixes

yavvorn committed:
Unittest Mocking done

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -46,8 +46,9 @@ def create_expense():
         query = "INSERT INTO expenses (expense, value) VALUES (%s, %s)"
         query_executor(query, (expense_name, expense_value), get_result=False)
         return {}, 201
```
return created expense as response

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -46,8 +46,9 @@ def create_expense():
+
     except Exception as e:
-        return {"error": "Failed to create expense"}, 500
+        return {"error": "Failed to create expense"}, 400
```
return 400 only on invalid data

slavovthinks replied:
validate:
```python
 expense_name = data.get('expense')
 expense_value = data.get('value')
```

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -59,30 +60,31 @@ def update_expense():
+        expense = query_executor(query, (expense_name, expense_value, expense_id), get_result=False)
+        if not expense:
+            return {"error": "Expense doesn't exist"}, 404
+        return {}, 204
```
Same validation 💩

slavovthinks replied:
use 200 (OK)

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -59,30 +60,31 @@ def update_expense():
 
-    return expense
+    except Exception as e:
+        return {"error": "Expense not found."}, 404
```
on
```
if __name__ == '__main__':

    
          
            
    

          
          Expand Down
    
    
  
    app.run(debug=True)
    ```
    Move DEBUG to an env var maybe APP_DEBUG_MODE or idk

slavovthinks commented on tests/test_app.py:
```diff
@@ -18,69 +17,60 @@ def test_get_expense_existing(mocker):
+
+def test_update_expense_happy_path(mocker):
+    update_expense_data = {"id": 123, "expense": "Groceries", "value": 150}
+    query_executor_mock = Mock(return_value=True)  # Simulating a successful update
```
Query executor never returns True bruh

slavovthinks reviewed (changes requested):
your turn bruh

yavvorn committed:
validate_create_expense created and create_expense modified

slavovthinks commented on spending_tracker/validators.py:
```diff
@@ -0,0 +1,11 @@
+    expense_value = payload.get('value')
+
+    is_expense_valid = expense_name and isinstance(expense_name, str)
+    is_value_valid = isinstance(expense_value, (int, float)) and expense_value > 0  # isnumeric doesn't work on JSON
```
Because `isnumeric` is a `string` method https://www.w3schools.com/python/ref_string_isnumeric.asp

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -39,16 +40,19 @@ def create_expense():
 
-    expense_name = data.get('expense')
-    expense_value = data.get('value')
+    if not data:
```
move that inside `validate_create_expense`

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -39,16 +40,19 @@ def create_expense():
+        return {"error": "Invalid input: No data provided"}, 400
 
-    try:
+    checker = validate_create_expense(data)
```
1. a better variable name would be `is_data_valid` its a boolean variable. Its a good practice to name boolean variable `is_{whatever-True-means}`.
2. We dont need variables at all. It would be best if we do the validation + error handling in the beginning of the function and then continue with its actual body. 
```
if not validate_create_expense(data):
    return {"error": "invalid stuff bruf"}, 400

slavovthinks committed:
Fix some stuff

yavvorn committed:
updated unittests, removed .lower() due to error

slavovthinks commented on spending_tracker/app.py:
```diff
@@ -89,4 +89,4 @@ def delete_expense(expense_id):
 
 
 if __name__ == '__main__':
-    app.run(debug=os.getenv("ENVIRONMENT").lower() == "dev")
```
Revert. The issue is caused by ENVIRONMENT env var not being set

slavovthinks commented on spending_tracker/validators.py:
```diff
@@ -6,7 +6,7 @@ def validate_create_expense(payload):
 
     is_expense_valid = expense_name and isinstance(expense_name, str)
-    is_value_valid = isinstance(expense_value, (int, float)) and expense_value > 0  # isnumeric doesn't work on JSON
+    is_value_valid = expense_value and isinstance(expense_value, (int, float)) and expense_value > 0
```
expense_value check is redundant as isinstance(expense_value, (int, float)) will be False if it is None

slavovthinks committed:
Pak zapovqdai

yavvorn committed:
unittests have been updated, get requests return id, peleze have mercy lord radostin-chan

yavvorn committed:
new db added for users, base user funcs added

yavvorn committed:
added validators for email, password and pass hashing

yavvorn committed:
created routes and added users/expenses there, added extensions for bcrypt

yavvorn committed:
started fixing unittests

yavvorn committed:
added jwt auth as well as env var for it, moved tests folder to spending_tracker for the imports

yavvorn committed:
removed unnecessary auth prefix to login route

yavvorn committed:
fix mocker patch paths for tests to work

yavvorn committed:
added unit tests for users, updated users update func

yavvorn committed:
added login that cross-ref info in DB and generates JWT token if logged in

yavvorn committed:
added a /me page

slavovthinks committed:
Welcome again MF

yavvorn committed:
add jwt token to all tests where it is required, change home route to return all expenses of the logged in user, comment out /me for now

slavovthinks reviewed (approved):
LGTM!
